libraries/ms-common/src/can_datagram.c

```diff
@@ -0,0 +1,476 @@
+    } else {
+      memcpy(&dt->data_len, rx_data, DATA_LEN_SIZE_BYTES);
+      event_raise_no_data(DATAGRAM_EVENT_DATA);
+      LOG_DEBUG("data\n");
+    }
+  }
+}
+
+static void prv_process_data(Fsm *fsm, const Event *e, void *context) {
+  CanDatagram *dt = context;
+  uint16_t data_len = dt->data_len;
+  if (s_store.mode == CAN_DATAGRAM_MODE_TX) {
+    uint8_t data_bytes_to_send = 0;
+    // TX max of 4 messages at a time
+    for (int msg = 0; msg < MAX_CAN_TX; msg++) {
+      data_bytes_to_send = (data_len - s_store.tx_bytes_sent < CAN_BUFFER_SIZE)
+                               ? (data_len - s_store.tx_bytes_sent)
+                               : CAN_BUFFER_SIZE;
+      memcpy(s_can_buffer, dt->data + s_store.tx_bytes_sent, (size_t)data_bytes_to_send);
+      s_store.tx_bytes_sent += data_bytes_to_send;
+      s_store.tx_cb(s_can_buffer, data_bytes_to_send, false);
+      if (s_store.tx_bytes_sent == data_len) {
+        event_raise_no_data(DATAGRAM_EVENT_COMPLETE);
+        s_store.tx_bytes_sent = 0;
+        LOG_DEBUG("datagram complete!\n");
+        watchdog_cancel(&s_watchdog);
+        return;
+      }
+    }
+    event_raise_no_data(DATAGRAM_EVENT_DATA);
+  } else {
+    uint8_t *rx_data;
+    size_t len;
+    if (!prv_read_rx_buffer(&rx_data, &len)) {
+      event_raise_no_data(DATAGRAM_EVENT_DATA);
+    } else {
+      memcpy(dt->data + s_store.rx_bytes_read, rx_data, len);
+      s_store.rx_bytes_read += len;
+      if (s_store.rx_bytes_read >= dt->data_len) {
+        s_store.rx_bytes_read = 0;
+        if (prv_can_datagram_compute_crc() == dt->crc) {
+          event_raise_no_data(DATAGRAM_EVENT_COMPLETE);
+          LOG_DEBUG("datagram complete!\n");
+        } else {
+          event_raise_no_data(DATAGRAM_EVENT_ERROR);
+          LOG_DEBUG("CRC VALUE DID NOT MATCH. EXITING WITH ERROR...\n");
+        }
+      } else {
+        event_raise_no_data(DATAGRAM_EVENT_DATA);
+      }
+    }
+  }
+}
+
+static void prv_process_done(Fsm *fsm, const Event *e, void *context) {
+  if (e->id == DATAGRAM_EVENT_COMPLETE) {
+    s_store.status = DATAGRAM_STATUS_COMPLETE;
+  } else {
+    s_store.status = DATAGRAM_STATUS_ERROR;
+  }
+  watchdog_cancel(&s_watchdog);
+}
+
+static void prv_init_fsm(void *context) {
+  fsm_init(&s_dt_fsm, "can_dt_fsm", &state_idle, context);
+  fsm_state_init(state_protocol_version, prv_process_protocol_version);
+  fsm_state_init(state_crc, prv_process_crc);
+  fsm_state_init(state_dt_type, prv_process_dt_type);
+  fsm_state_init(state_dst_len, prv_process_dst_len);
+  fsm_state_init(state_dst, prv_process_dst);
+  fsm_state_init(state_data_len, prv_process_data_len);
+  fsm_state_init(state_data, prv_process_data);
+  fsm_state_init(state_done, prv_process_done);
+}
+
+static void prv_rx_watchdog_expiry_cb(void *context) {
+  event_raise_no_data(DATAGRAM_EVENT_ERROR);
+  LOG_DEBUG("WATCHDOG EXPIRED. EXITING WITH ERROR...\n");
+}
+
+StatusCode can_datagram_init(CanDatagramSettings *settings) {
```

Comment:
The API for this module is a little weird - to send/receive multiple messages you'd have to reinitialize it! Here is my idea for how to make the API more intuitive:
- make `can_datagram_init` a thin init function that only initializes the fsm and status and stuff like that
- give `state_done` a way to return to `state_idle` once we're done rxing/txing a datagram
- have `can_datagram_start_listening` and `can_datagram_start_tx` functions which can only be called in `state_idle`, and which start an RX or TX respectively
  - `can_datagram_start_listening` initializes everything for RX mode, and takes the data buffer where we put the datagram data
  - `can_datagram_start_tx` initializes everything for TX mode, takes all the datagram fields, and starts the TX
- also give users a way to know if we're currently in the middle of a tx/rx so they know if it's ok to start listening/txing
- set the event data in all the events (or at least the completion event) to the state we're in (tx/rx) so that downstream clients know if we've just txed or rxed
- document all this in the header!

Reply:
- [ ] make can_datagram_init a thin init function that only initializes the fsm and status and stuff like that

- [ ] give state_done a way to return to state_idle once we're done rxing/txing a datagram

- [ ] have can_datagram_start_listening and can_datagram_start_tx functions which can only be called in state_idle, and which start an RX or TX respectively

- [ ] can_datagram_start_listening initializes everything for RX mode, and takes the data buffer where we put the datagram data
    - Now using tx and rx config structs  

- [ ] can_datagram_start_tx initializes everything for TX mode, takes all the datagram fields, and starts the TX

also give users a way to know if we're currently in the middle of a tx/rx so they know if it's ok to start listening/txing
set the event data in all the events (or at least the completion event) to the state we're in (tx/rx) so that downstream clients know if we've just txed or rxed
   - there are now two different statuses used for tx/rx completion